test/framework/clusterctl/client.go

```diff
@@ -96,24 +88,33 @@ func Init(_ context.Context, input InitInput) {
 
 // InitWithBinary uses clusterctl binary to run init with the list of providers defined in the local repository.
 func InitWithBinary(_ context.Context, binary string, input InitInput) {
-	log.Logf("clusterctl init --core %s --bootstrap %s --control-plane %s --infrastructure %s --ipam %s --runtime-extension %s --config %s --kubeconfig %s",
-		input.CoreProvider,
-		strings.Join(input.BootstrapProviders, ","),
-		strings.Join(input.ControlPlaneProviders, ","),
-		strings.Join(input.InfrastructureProviders, ","),
-		strings.Join(input.IPAMProviders, ","),
-		strings.Join(input.RuntimeExtensionProviders, ","),
-		input.ClusterctlConfigPath,
-		input.KubeconfigPath,
-	)
+	args := calculateClusterCtlInitArgs(input)
+	log.Logf("clusterctl %s", strings.Join(args, ""))
+
+	cmd := exec.Command(binary, args...) //nolint:gosec // We don't care about command injection here.
+
+	out, err := cmd.CombinedOutput()
+	_ = os.WriteFile(filepath.Join(input.LogFolder, "clusterctl-init.log"), out, 0644) //nolint:gosec // this is a log file to be shared via prow artifacts
+	var stdErr string
+	if err != nil {
+		if exitErr, ok := err.(*exec.ExitError); ok {
+			stdErr = string(exitErr.Stderr)
+		}
+	}
+	Expect(err).ToNot(HaveOccurred(), "failed to run clusterctl init:\nstdout:\n%s\nstderr:\n%s", string(out), stdErr)
+}
 
+func calculateClusterCtlInitArgs(input InitInput) []string {
 	args := []string{"init", "--config", input.ClusterctlConfigPath, "--kubeconfig", input.KubeconfigPath}
 	if input.CoreProvider != "" {
 		args = append(args, "--core", input.CoreProvider)
 	}
 	if len(input.BootstrapProviders) > 0 {
 		args = append(args, "--bootstrap", strings.Join(input.BootstrapProviders, ","))
 	}
+	if len(input.ControlPlaneProviders) > 0 {
+		args = append(args, "--control-plane", strings.Join(input.ControlPlaneProviders, ","))
+	}
```

Comment:
Previously we did not set `--control-plane` in InitWithBinary.

The result for the v1.2 => main test was that the following providers where installed during init:
```
 k get provider -A
NAMESPACE                           NAME                    AGE   TYPE                     PROVIDER      VERSION
capd-system                         infrastructure-docker   8s    InfrastructureProvider   docker        v1.2.8
capi-kubeadm-bootstrap-system       bootstrap-kubeadm       9s    BootstrapProvider        kubeadm       v1.2.8
capi-kubeadm-control-plane-system   control-plane-kubeadm   8s    ControlPlaneProvider     kubeadm       v1.4.99
capi-system                         cluster-api             9s    CoreProvider             cluster-api   v1.2.8
```

Reply:
This only seems to be an issue if there are newer versions of the same contract, e.g. the v0.4 => main test was okay:
```
k get provider -A
NAMESPACE                           NAME                    AGE   TYPE                     PROVIDER      VERSION
capd-system                         infrastructure-docker   26s   InfrastructureProvider   docker        v0.4.8
capi-kubeadm-bootstrap-system       bootstrap-kubeadm       26s   BootstrapProvider        kubeadm       v0.4.8
capi-kubeadm-control-plane-system   control-plane-kubeadm   26s   ControlPlaneProvider     kubeadm       v0.4.8
capi-system                         cluster-api             26s   CoreProvider             cluster-api   v0.4.8
```